snooty/postprocess.py

```diff
@@ -532,11 +533,34 @@ class TabsSelectorHandler(Handler):
     def __init__(self, context: Context) -> None:
         super().__init__(context)
         self.selectors: Dict[str, List[Dict[str, MutableSequence[n.Text]]]] = {}
+        self.stack: List[str] = []
+        self.scanned_pattern: List[str] = []
+
+    def scan_for_pattern(self, fileid_stack: FileIdStack, node: n.Node) -> None:
+        starting_point = 0
+        target_pattern = ["tabs", "tabs", "procedure"]
+        target_pattern_len = len(target_pattern)
+        if len(self.scanned_pattern) > 0:
+            reverse_list = reversed(self.scanned_pattern)
+            for item in reverse_list:
+                if item == target_pattern[starting_point]:
+                    starting_point += 1
+                if starting_point >= target_pattern_len:
+                    self.context.diagnostics[fileid_stack.current].append(
+                        InvalidNestedTabStructure(
+                            " ".join(reversed(self.scanned_pattern)), node.start[0]
+                        )
+                    )
+                    return
 
     def enter_node(self, fileid_stack: FileIdStack, node: n.Node) -> None:
         if not isinstance(node, n.Directive):
             return
 
+        self.scan_for_pattern(fileid_stack, node)
```

Comment:
We might also want to scan for the pattern after exiting the node or after `self.scanned_pattern` adds the `procedure` directive, since `self.scanned_pattern` would only contain the `procedure` directive after the node has been exited and a new node is entered. (i.e. I think if the pattern exists at the end of a file, the pattern won't be caught by the scan since there's no more nodes to enter, but please correct me if I'm wrong!)

Reply:
@rayangler that is fair, yeah I believe you are right, the keyword being at the end of the file, because during `exit_page` we clear the list, so the pattern won't be caught. Good call.